cowprotocol/user_segmentation/user_segmentation_4872780.sql

```diff
@@ -0,0 +1,122 @@
+-- It is parametrised by chain and project.
+-- Status descriptions:
+-- Loyal - only traded {{project}} during calendar month
+-- Opportunistic - traded both {{project}} during calendar month and other projects
+-- Gone - was created before date of interest but didn't trade in given calendar month
+-- Didnt exist - user was created this month (this status is only possible for prev_status column)
+--
+-- Parameters:
+--  {{start_time}} - the trade timestamp for which the analysis should start (inclusive)
+--  {{chain_name}} - chain to do the analysis for
+--  {{project}} - project to do the analysis for
+--  {{current_status}} - filter to limit analysis to users that had this status as current for given calendar month
+--  {{previous_status}} - filter to limit analysis to users that had this status as previous for given calendar month
+
+
+with creation_month_per_user as (
+    -- TODO: use both dex and dex-agg here
+    select
+        taker as user_address,
+        min(date_trunc('Month', block_date)) as creation_month
+    from
+        dex_aggregator.trades
+    where
+        blockchain = '{{chain_name}}'
+    group by 1
+),
+
+all_active_users as (
+    -- TODO: use layered txs here to prevent miss-information
+    select distinct
+        taker as user_address,
+        date_trunc('Month', block_date) as block_month,
+        case when project in ('1inch', '1inch Limit Order Protocol') then '1inch' else project end as project
+    from
+        dex_aggregator.trades
+    where
+        blockchain = '{{chain_name}}'
+        and
+        block_date >= timestamp '{{start_time}}' - interval '1' month
+),
+
+users_backbone as (
+    select
+        user_address,
+        a.dt as block_month,
+        null as project
+    from
+        all_active_users
+    cross join unnest(sequence(date_trunc('Month', timestamp '{{start_time}}') - interval '1' month, current_date, interval '1' month)) as a (dt)
+),
+
+all_users as (
+    select *
+    from
+        users_backbone
+
+    union all
+
+    select *
+    from
+        all_active_users
+),
+
+user_w_creation_date as (
+    select
+        all_users.user_address,
+        block_month,
+        creation_month,
+        project
+    from
+        all_users
+    left join creation_month_per_user on all_users.user_address = creation_month_per_user.user_address
+    where
+        creation_month <= block_month
+),
+
+users_labeled as (
+    select
+        user_address,
+        block_month,
+        any_value(creation_month) as creation_month,
+        case
+            when count(distinct project) = 1 and max(project) = '{{project}}' then 'Loyal'
+            when contains(array_agg(project) filter (where project is not null), '{{project}}') then 'Opportunistic'
+            when count(distinct project) >= 1 and not contains(array_agg(project) filter (where project is not null), '{{project}}') then 'Lost'
+            else 'Gone'
+        end as user_type
+    from
+        user_w_creation_date
```

Comment:
nit: could join `creation_month_per_user` here to save one CTE and make the query a bit shorter